### PR TITLE
br: error log optimization

### LINKED ISSUE #27015: backup failed summary should be quieter
## Enhancement

The current version of backup failed summary is very verbose: it prints each ranges failed to be backed up, with stack trace of the fail reason, even the reason is trivially 'context cancelled'. for example:
`[2021/08/09 02:14:32.193 +00:00] [INFO] [collector.go:195] ["Full backup failed summary"] [total-ranges=8533] [ranges-succeed=3332] [ranges-failed=5201] [backup-total-ranges=8533] [unit-name="range start:? end:?"] [error="rpc error: code = Canceled desc = context canceled"] [errorVerbose="rpc error: code = Canceled desc = context canceled\ngithub.com/tikv/pd/client.(*client).GetAllStores\n\tgithub.com/tikv/pd@v0.0.0-20210105112549-e5be7fd38659/client/client.go:657\ngithub.com/pingcap/br/pkg/conn.GetAllTiKVStores\n\tgithub.com/pingcap/br/pkg/conn/conn.go:78\ngithub.com/pingcap/br/pkg/backup.(*Client).BackupRange\n\tgithub.com/pingcap/br/pkg/backup/client.go:524\ngithub.com/pingcap/br/pkg/backup.(*Client).BackupRanges.func2.1\n\tgithub.com/pingcap/br/pkg/backup/client.go:472\ngithub.com/pingcap/br/pkg/utils.(*WorkerPool).ApplyOnErrorGroup.func1\n\tgithub.com/pingcap/br/pkg/utils/worker.go:73\ngolang.org/x/sync/errgroup.(*Group).Go.func1\n\tgolang.org/x/sync@v0.0.0-20201020160332-67f06af15bc9/errgroup/errgroup.go:57\nruntime.goexit\n\truntime/asm_amd64.s:1371"] [unit-name="range start:? end:?"] [error="rpc error: code = Canceled desc = context canceled"] [errorVerbose="rpc error: code = Canceled desc = context canceled\ngithub.com/tikv/pd/client.(*client).GetAllStores\n\tgithub.com/tikv/pd@v0.0.0-20210105112549-e5be7fd38659/client/client.go:657\ngithub.com/pingcap/br/pkg/conn.GetAllTiKVStores\n\tgithub.com/pingcap/br/pkg/conn/conn.go:78\ngithub.com/pingcap/br/pkg/backup.(*Client).BackupRange\n\tgithub.com/pingcap/br/pkg/backup/client.go:524\ngithub.com/pingcap/br/pkg/backup.(*Client).BackupRanges.func2.1\n\tgithub.com/pingcap/br/pkg/backup/client.go:472\ngithub.com/pingcap/br/pkg/utils.(*WorkerPool).ApplyOnErrorGroup.func1\n\tgithub.com/pingcap/br/pkg/utils/worker.go:73\ngolang.org/x/sync/errgroup.(*Group).Go.func1\n\tgolang.org/x/sync@v0.0.0-20201020160332-67f06af15bc9/errgroup/errgroup.go:57\nruntime.goexit\n\truntime/asm_amd64.s:1371"]
... (more than 80 replicates of the above omitted)`

It is annoying and also shadows information really useful. We can make it more friendly:

1. We can print each failed unit and its error in separated logs, the summary should be a simple summary. Line-based unix toolchains like `grep` could take good from this.
2. Stacktrace of trivial errors (including `context.Canceled` or `err if status.Code(err) == codes.Canceled`) shouldn't be printed. If (1) has been implemented, units failed due to these errors could be printed as lower level(say, don't print them to the terminal, or use the DEBUG level?).

A better failure log may seem like:
```
[INFO] ["Full backup failed summary"] [total-ranges=8533] [ranges-succeed=3332] [ranges-failed=5201] [backup-total-ranges=8533]
[ERROR] ["Failed unit information"] [unit-name="range start:? end:?"] [error="[BR:KV:ErrKVStorage]tikv storage occur I/O error"] [errorVerbose="[BR:KV:ErrKVStorage]tikv storage occur I/O error\ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0.20201126102027-b0a155152ca3/errors.go:174\ngithub.com/pingcap/errors.Trace\n\tgithub.com/pingcap/errors@v0.11.5-0.20201126102027-b0a155152ca3/juju_adaptor.go:15\ngithub.com/pingcap/br/pkg/backup.(*Client).BackupRange\n\tgithub.com/pingcap/br/pkg/backup/client.go:538\ngithub.com/pingcap/br/pkg/backup.(*Client).BackupRanges.func2.1\n\tgithub.com/pingcap/br/pkg/backup/client.go:472\ngithub.com/pingcap/br/pkg/utils.(*WorkerPool).ApplyOnErrorGroup.func1\n\tgithub.com/pingcap/br/pkg/utils/worker.go:73\ngolang.org/x/sync/errgroup.(*Group).Go.func1\n\tgolang.org/x/sync@v0.0.0-20201020160332-67f06af15bc9/errgroup/errgroup.go:57\nruntime.goexit\n\truntime/asm_amd64.s:1371"]
[INFO] ["Canceled units"] [units="[?, ..., ?]"]
 ```

### PR DESCRIPTION
**Pull request by fengou1**
### What problem does this PR solve?

Issue Number: close #27015

Problem Summary:
BackupRange is parallel work in a error group, once an error happened with the exhausted retry, all others coroutines been canceled, those cancelled work append stack traces are useless.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: br/pkg/backup/client.go [hunk 1]
```diff
@@ -452,7 +452,14 @@ func (bc *Client) BackupRanges(
...
 			err := bc.BackupRange(elctx, sk, ek, req, metaWriter, progressCallBack)
 			if err != nil {
-				return errors.Trace(err)
+				// The error due to context cancel, stack trace is meaningless, the stack shall be suspended (also clear)
+				if errors.Cause(err) == context.Canceled {
+					errors.SuspendStack(err)
+					return err
+				} else {
+					return errors.Trace(err)
+				}
```
Comment (kennytm) [inline br/pkg/backup/client.go:461]: i feel like this is something that should be handled in `errors.Trace` not here 🤔 

or we should change some `zap.Error` to `log.ShortError`...

Comment (fengou1) [reply br/pkg/backup/client.go:461]: errors.Trace comes from another remote package, put it into errors.Trace will has impact other repo, I guess.

Zap.Error and ShortError may not suitable for a real error, which we need call stack for investigation

File: br/pkg/backup/client.go [hunk 2]
```diff
@@ -452,7 +452,14 @@ func (bc *Client) BackupRanges(
 			elctx := logutil.ContextWithField(ectx, logutil.RedactAny("range-sn", id))
 			err := bc.BackupRange(elctx, sk, ek, req, metaWriter, progressCallBack)
 			if err != nil {
-				return errors.Trace(err)
+				// The error due to context cancel, stack trace is meaningless, the stack shall be suspended (also clear)
+				if errors.Cause(err) == context.Canceled {
+					errors.SuspendStack(err)
+					return err
```
Comment (YuJuncen) [inline br/pkg/backup/client.go:458]:
```suggestion
					return errors.SuspendStack(err)
```

**Review by YuJuncen — COMMENTED** [review]
rest LGTM.

BTW, Other places may also contains `context.Canceled`... I agree with @kennytm, maybe we'd better replace:

https://github.com/pingcap/tidb/blob/fec2938c1379270bf9939822c1abfe3d7244c174/br/pkg/summary/collector.go#L191-L193

With something like 
```go
if errors.Cause(reason) != context.Canceled {
 	logFields = append(logFields, zap.String("unit-name", unitName), zap.Error(reason)) 
} else {
        // collect all canceled units, then maybe print them at another log.
       canceledUnits = append(canceledUnits, unitName)
}
// ...
log.Info("units canceled", zap.Strings(canceledUnits))
```

File: br/pkg/summary/collector.go [hunk 3]
```diff
@@ -188,9 +190,16 @@ func (tc *logCollector) Summary(name string) {
 	}
 
 	if len(tc.failureReasons) != 0 || !tc.successStatus {
+		canceledUnits := make([]string, 0)
 		for unitName, reason := range tc.failureReasons {
-			logFields = append(logFields, zap.String("unit-name", unitName), zap.Error(reason))
+			if berror.Cause(reason) != context.Canceled {
+				logFields = append(logFields, zap.String("unit-name", unitName), zap.Error(reason))
+			} else {
+				canceledUnits = append(canceledUnits, unitName)
```
Comment (YuJuncen) [inline br/pkg/summary/collector.go:198]: If only log number, maybe can make `canceledUnits` a integer?

**Review by YuJuncen — APPROVED** [review]
rest lgtm

File: br/pkg/summary/collector.go [hunk 4]
```diff
@@ -188,9 +190,16 @@ func (tc *logCollector) Summary(name string) {
 	}
 
 	if len(tc.failureReasons) != 0 || !tc.successStatus {
+		var canceledUnits int
 		for unitName, reason := range tc.failureReasons {
-			logFields = append(logFields, zap.String("unit-name", unitName), zap.Error(reason))
+			if berror.Cause(reason) != context.Canceled {
+				logFields = append(logFields, zap.String("unit-name", unitName), zap.Error(reason))
+			} else {
+				canceledUnits = canceledUnits + 1
```
Comment (kennytm) [inline br/pkg/summary/collector.go:198]:
```suggestion
				canceledUnits++
```

**Review by kennytm — APPROVED** [review]
rest lgtm

File: br/pkg/summary/collector.go (hunk 4 above)
Comment (fengou1) [reply br/pkg/summary/collector.go]: fixed

**Pull request merged**